packages/flutter/lib/src/cupertino/context_menu.dart

```diff
@@ -0,0 +1,1099 @@
+
+  /// The callback to call when tapping on the child when the ContextMenu is
+  /// open.
+  final VoidCallback onTap;
+
+  /// The widget to show when the ContextMenu is open.
+  ///
+  /// If not specified, [child] will be shown.
+  ///
+  /// This can be used to shown an entirely different widget than the preview,
+  /// but it can also show a slight variation of the child. As a simple example,
+  /// the child could be given rounded corners in the preview but have sharp
+  /// corners when in the page.
+  final Widget preview;
+
+  @override
+  _ContextMenuState createState() => _ContextMenuState();
+}
+
+class _ContextMenuState extends State<ContextMenu> with TickerProviderStateMixin {
+  final GlobalKey _childGlobalKey = GlobalKey();
+  AnimationController _decoyController;
+  Rect _decoyChildEndRect;
+
+  OverlayEntry _lastOverlayEntry;
+  double _childOpacity = 1.0;
+  _ContextMenuRoute<void> _route;
+
+  @override
+  void initState() {
+    super.initState();
+    _decoyController = AnimationController(
+      duration: const Duration(milliseconds: 600),
+      vsync: this,
+    );
+    _decoyController.addStatusListener(_onDecoyAnimationStatusChange);
+  }
+
+  // Determine the _ContextMenuOrientation based on the location of the original
+  // child in the screen.
+  _ContextMenuOrientation get _contextMenuOrientation {
+    final Rect childRect = _getRect(_childGlobalKey);
+    final double screenWidth = MediaQuery.of(context).size.width;
+
+    final double center = screenWidth / 2;
+    final bool centerDividesChild = childRect.left < center && childRect.right > center;
+    final double minCenterWidth = screenWidth / 4;
+    final double maxCenterWidth = screenWidth / 3;
+    final bool isCenterWidth = childRect.width >= minCenterWidth
+      && childRect.width <= maxCenterWidth;
+    if (centerDividesChild && isCenterWidth) {
+      return _ContextMenuOrientation.center;
+    }
+
+    if (childRect.center.dx > center) {
+      return _ContextMenuOrientation.right;
+    }
+
+    return _ContextMenuOrientation.left;
+  }
+
+  // Push the new route and open the ContextMenu overlay.
+  void _openContextMenu() {
+    HapticFeedback.lightImpact();
+
+    setState(() {
+      _childOpacity = 0.0;
+    });
+
+    _route = _ContextMenuRoute<void>(
+      actions: widget.actions,
+      barrierLabel: 'Dismiss',
+      filter: ui.ImageFilter.blur(
+        sigmaX: 5.0,
+        sigmaY: 5.0,
+      ),
+      onTap: widget.onTap,
+      contextMenuOrientation: _contextMenuOrientation,
+      previousChildRect: _decoyChildEndRect,
+      builder: (BuildContext context) {
+        return widget.preview ?? widget.child;
+      },
+    );
+    Navigator.of(context, rootNavigator: true).push<void>(_route);
+    _route.animation.addStatusListener(_routeAnimationStatusListener);
+  }
+
+  // The OverlayEntry that positions widget.child directly on top of its
+  // original position in this widget.
+  OverlayEntry get _overlayEntry {
```

Comment:
It feels a bit weird to have a getter that returns a new OverlayEntry every time it gets accessed.